### ISSUE TITLE
Multi-GPU Training on SLURM

### ISSUE THREAD

**Issue by farrosalferro** [autonomousvision/navsim#72]
Thank you for hosting the challenge this year. I have a question regarding training the model in a multi-gpu environment, specifically with slurm workload manager. I tried running this command in a node with 4 gpus and 96 threads:

#!/bin/bash
#SBATCH --job-name=transfuser_training
###########RESOURCES###########
#SBATCH --partition=48-4
#SBATCH --gres=gpu:4
#SBATCH --mem=240G
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --cpus-per-task=24
###############################
#SBATCH --output=logs/transfuser_training_ddp.out
#SBATCH --error=logs/transfuser_training_ddp.err
#SBATCH -v

source ~/miniconda3/etc/profile.d/conda.sh
conda activate navsim

TRAIN_TEST_SPLIT=navtrain

srun python $NAVSIM_DEVKIT_ROOT/navsim/planning/script/run_training.py \
agent=transfuser_agent \
experiment_name=training_transfuser_latent_agent \
train_test_split=$TRAIN_TEST_SPLIT \
worker=ray_distributed \
agent.config.latent=True \
trainer.params.max_epochs=1 \

However, the process stops after it loads the logs, even though it says "start training". Here is the output of .out and .err (after loading the log) files:

[2025-03-18 09:34:56,898][__main__][INFO] - Global Seed set to 0
[2025-03-18 09:34:56,901][__main__][INFO] - Path where all results are stored: /po1/farrosalferro/navsim/exp/training_transfuser_latent_agent/2025.03.18.09.34.47
[2025-03-18 09:34:56,901][__main__][INFO] - Building Agent
[2025-03-18 09:34:57,100][__main__][INFO] - Global Seed set to 0
[2025-03-18 09:34:57,103][__main__][INFO] - Path where all results are stored: /po1/farrosalferro/navsim/exp/training_transfuser_latent_agent/2025.03.18.09.34.47
[2025-03-18 09:34:57,103][__main__][INFO] - Building Agent
[2025-03-18 09:34:57,157][__main__][INFO] - Global Seed set to 0
[2025-03-18 09:34:57,158][__main__][INFO] - Path where all results are stored: /po1/farrosalferro/navsim/exp/training_transfuser_latent_agent/2025.03.18.09.34.48
[2025-03-18 09:34:57,159][__main__][INFO] - Building Agent
[2025-03-18 09:34:57,384][__main__][INFO] - Global Seed set to 0
[2025-03-18 09:34:57,387][__main__][INFO] - Path where all results are stored: /po1/farrosalferro/navsim/exp/training_transfuser_latent_agent/2025.03.18.09.34.48
[2025-03-18 09:34:57,387][__main__][INFO] - Building Agent
[2025-03-18 09:34:59,990][timm.models._builder][INFO] - Loading pretrained weights from Hugging Face hub (timm/resnet34.a1_in1k)
[2025-03-18 09:34:59,993][timm.models._builder][INFO] - Loading pretrained weights from Hugging Face hub (timm/resnet34.a1_in1k)
[2025-03-18 09:34:59,993][timm.models._builder][INFO] - Loading pretrained weights from Hugging Face hub (timm/resnet34.a1_in1k)
[2025-03-18 09:34:59,994][timm.models._builder][INFO] - Loading pretrained weights from Hugging Face hub (timm/resnet34.a1_in1k)
[2025-03-18 09:35:00,211][timm.models._hub][INFO] - [timm/resnet34.a1_in1k] Safe alternative available for 'pytorch_model.bin' (as 'model.safetensors'). Loading weights using safetensors.
[2025-03-18 09:35:00,211][timm.models._hub][INFO] - [timm/resnet34.a1_in1k] Safe alternative available for 'pytorch_model.bin' (as 'model.safetensors'). Loading weights using safetensors.
[2025-03-18 09:35:00,211][timm.models._hub][INFO] - [timm/resnet34.a1_in1k] Safe alternative available for 'pytorch_model.bin' (as 'model.safetensors'). Loading weights using safetensors.
[2025-03-18 09:35:00,211][timm.models._hub][INFO] - [timm/resnet34.a1_in1k] Safe alternative available for 'pytorch_model.bin' (as 'model.safetensors'). Loading weights using safetensors.
[2025-03-18 09:35:01,013][timm.models._builder][INFO] - Missing keys (fc.weight, fc.bias) discovered while loading pretrained weights. This is expected if model is being adapted.
[2025-03-18 09:35:01,016][timm.models._builder][INFO] - Missing keys (fc.weight, fc.bias) discovered while loading pretrained weights. This is expected if model is being adapted.
[2025-03-18 09:35:01,018][timm.models._builder][INFO] - Missing keys (fc.weight, fc.bias) discovered while loading pretrained weights. This is expected if model is being adapted.
[2025-03-18 09:35:01,018][timm.models._builder][INFO] - Missing keys (fc.weight, fc.bias) discovered while loading pretrained weights. This is expected if model is being adapted.
[2025-03-18 09:35:01,690][__main__][INFO] - Building Lightning Module
[2025-03-18 09:35:01,690][__main__][INFO] - Building SceneLoader
[2025-03-18 09:35:01,752][__main__][INFO] - Building Lightning Module
[2025-03-18 09:35:01,752][__main__][INFO] - Building SceneLoader
[2025-03-18 09:35:01,758][__main__][INFO] - Building Lightning Module
[2025-03-18 09:35:01,759][__main__][INFO] - Building SceneLoader
[2025-03-18 09:35:01,759][__main__][INFO] - Building Lightning Module
[2025-03-18 09:35:01,759][__main__][INFO] - Building SceneLoader
[2025-03-18 09:38:35,623][__main__][INFO] - Building Datasets
[2025-03-18 09:38:35,623][__main__][INFO] - Building Datasets
[2025-03-18 09:38:35,623][__main__][INFO] - Building Datasets
[2025-03-18 09:38:35,624][__main__][INFO] - Building Datasets
[2025-03-18 09:38:35,648][__main__][INFO] - Num training samples: 85109
[2025-03-18 09:38:35,649][__main__][INFO] - Num training samples: 85109
[2025-03-18 09:38:35,649][__main__][INFO] - Num training samples: 85109
[2025-03-18 09:38:35,649][__main__][INFO] - Num training samples: 85109
[2025-03-18 09:38:35,650][__main__][INFO] - Num validation samples: 18179
[2025-03-18 09:38:35,650][__main__][INFO] - Building Trainer
[2025-03-18 09:38:35,650][__main__][INFO] - Num validation samples: 18179
[2025-03-18 09:38:35,650][__main__][INFO] - Building Trainer
[2025-03-18 09:38:35,651][__main__][INFO] - Num validation samples: 18179
[2025-03-18 09:38:35,651][__main__][INFO] - Building Trainer
[2025-03-18 09:38:35,651][__main__][INFO] - Num validation samples: 18179
[2025-03-18 09:38:35,651][__main__][INFO] - Building Trainer
[2025-03-18 09:38:35,781][__main__][INFO] - Starting Training
[2025-03-18 09:38:35,781][__main__][INFO] - Starting Training
[2025-03-18 09:38:35,781][__main__][INFO] - Starting Training
[2025-03-18 09:38:35,782][__main__][INFO] - Starting Training

Using 16bit Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Trainer(limit_train_batches=1.0) was configured so 100% of the batches per epoch will be used..
Trainer(limit_val_batches=1.0) was configured so 100% of the batches will be used..
Trainer(val_check_interval=1.0) was configured so validation will run at the end of the training epoch..
You are using a CUDA device ('NVIDIA RTX A6000') that has Tensor Cores. To properly utilize them, you should set torch.set_float32_matmul_precision('medium' | 'high') which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
[rank: 0] Seed set to 0
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/4
[rank: 3] Seed set to 0
Initializing distributed: GLOBAL_RANK: 3, MEMBER: 4/4
[rank: 1] Seed set to 0
Initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/4
[rank: 2] Seed set to 0
Initializing distributed: GLOBAL_RANK: 2, MEMBER: 3/4
distributed_backend=nccl
All distributed processes registered. Starting with 4 processes

Any help would be appreciated, thank you.

**Comment by DanielDauner** [comment]
Hi @farrosalferro 

I would expect the problem lies in the `worker=ray_distributed`. The `ray` library is often causing issues on slurm and I was never able to reliably fix that on our cluster. A workaround is to use `worker=single_machine_thread_pool` during training on slurm.

We also recommend to first run dataset caching (ie. locally) for stability. Here, using the `ray_distributed` argument will make a huge speed improvement compared to `single_machine_thread_pool`. After dataset caching, you can run training on the cluster (where you only need to store the dataset cache). When training with a dataset cache, using `single_machine_thread_pool` does not slow down training. 

Best,
Daniel

**Comment by DanielDauner** [comment]
Maybe this issue is helpful (https://github.com/autonomousvision/navsim/issues/13)
